codec.go

```diff
@@ -45,11 +44,77 @@ func FromBytesOrNil(input []byte) UUID {
 	return uuid
 }
 
+var errInvalidFormat = errors.New("uuid: invalid UUID format")
+
+func fromHexChar(c byte) byte {
+	switch {
+	case '0' <= c && c <= '9':
+		return c - '0'
+	case 'a' <= c && c <= 'f':
+		return c - 'a' + 10
+	case 'A' <= c && c <= 'F':
+		return c - 'A' + 10
+	}
+	return 255
+}
+
+// Parse parses the UUID stored in the string text. Parsing and supported
+// formats are the same as UnmarshalText.
+func (u *UUID) Parse(s string) error {
```

Comment:
Why not make `Parse()` delegate to `UnmarshalText()` or vice versa?  The conversion from `string` to `[]byte` or `[]byte` to `string` is almost always optimized out by the compiler and this is non-trivial code to be duplicating.

Reply:
> The conversion from string to []byte or []byte to string is almost always optimized out by the compiler and this is non-trivial code to be duplicating.

In some cases this is true and the runtime can stack allocate the string so only needs to do a memmove (see: [`slicebytetostring()`](https://github.com/golang/go/blob/9f0234214473dfb785a5ad84a8fc62a6a395cbc3/src/runtime/string.go#L81-L116)), but that is not true here since `([]byte)(s)` escapes to heap (you can check this by making the suggested change and running `go build -i -gcflags='-m'`).

Reply:
Additionally, delegating to `UnmarshalText` could be DOS vector if this was used to parse untrusted input (a malicious actor could pass very large "UUID" causing the application to allocate large amounts of memory).

Reply:
> My concern is that this is a fairly complex block of code that's identical to UnmarshalText() other than one processes string and the other []byte. It would be better in the long run to not have the duplication. 

I don't like the duplication either, but this is the only way to do this without allocating. The fact that we have 100% test coverage considerably alleviates these concerns IMO. In the long run this could be replaced by a generic function that takes both strings and byte slices, but that should only be done once Go 1.17 is no longer in wide use.

> What's the result if UnmarshalText() delegates to Parse() instead?

Same result and same issues around creating a copy of untrusted input.

> Also, since the logic is the same, any DOS vector in UnmarshalText() exists here as well. 

The DOS vector only exists if we delegate the call to `UnmarshalText/Parse` from `Parse/UnmarshalText` and have to allocate to create a copy of untrusted input.